# Prepare EEG data for training of machine-learning models
+ Import data.
+ Apply filters (bandpass).
+ Detect potential bad channels and replace them by interpolation.
+ Detect potential bad epochs and remove them.

## Import packages & links

In [1]:
# Import packages
import os
import sys
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
sys.path.insert(0, os.path.dirname(os.getcwd()))

import mne
#%matplotlib inline
#from mayavi import mlab

In [3]:
from config import ROOT, PATH_CODE, PATH_DATA, PATH_OUTPUT, PATH_METADATA

In [4]:
filename_labels = PATH_METADATA + "Screening_children5a_summary_new.txt" 
metadata = pd.read_csv(filename_labels, sep='\t')
metadata.head()

id_child groupDDP atRiskOrNotDDP dyslexicAtMidGroup3DDP  \
0      001        4        unclear                      1   
1      002  missing        missing                missing   
2      003    3Ctrl      notAtRisk                      0   
3      004  missing        missing                missing   
4      005    3Ctrl      notAtRisk                      0   

               assignment1              assignment2              assignment3  \
0            notEnoughInfo            notEnoughInfo            notEnoughInfo   
1  notAtRisk_highestScores  notAtRisk_highestScores  notAtRisk_highestScores   
2  notAtRisk_highestScores  notAtRisk_highestScores  notAtRisk_highestScores   
3  notAtRisk_highestScores  notAtRisk_highestScores  notAtRisk_highestScores   
4  notAtRisk_highestScores  notAtRisk_highestScores  notAtRisk_highestScores   

               assignment4  childInfoPresent  relativeInfoPresent  mmr_2mth  \
0            notEnoughInfo                 1                    1         1   
1  notAtRisk_highestScores                 0                    1         0   
2  notAtRisk_highestScores                 1                    1         1   
3  notAtRisk_highestScores                 1                    1         1   
4  notAtRisk_highestScores                 1                    1         1   

   mmr_5mth  mmr_11mth  mmr_17mth  mmr_23mth  mmr_29mth  mmr_35mth  mmr_41mth  \
0         0          1          1          1          1          1          1   
1         0          0          0          1          1          1          0   
2         0          1          0          1          0          1          1   
3         1          1          0          1          1          0          1   
4         0          1          1          0          1          1          1   

   mmr_47mth  
0          1  
1          0  
2          0  
3          0  
4          0

In [5]:
metadata.shape

(336, 19)

## Search all *.cnt files and check for how many we have a label

In [4]:
PATH_DATA_01 = os.path.join(PATH_DATA, "17mnd mmn")

import fnmatch
import warnings
warnings.filterwarnings('ignore')

import helper_functions

dirs = os.listdir(PATH_DATA_01)
cnt_files = fnmatch.filter(dirs, "*.cnt")

In [5]:
found_ids = [x[:3] for x in cnt_files]
idx = np.where(np.array(found_ids) == '036')[0]
[cnt_files[x] for x in idx]

['036_17_mc_mmn25_wk.cnt', '036_17_mc_mmn36.cnt']

In [6]:
metadata[metadata['id_child'] == '036']['groupDDP'].values[0]

'3Ctrl'

In [7]:
labels = []

found_ids = [x[:3] for x in cnt_files]
for ID in list(set(found_ids)): 
    idx = np.where(np.array(found_ids) == ID)[0]
    filenames = [cnt_files[x] for x in idx]
    label = metadata[metadata['id_child'] == ID]['groupDDP'].values[0]
    label_risk = metadata[metadata['id_child'] == ID]['assignment4'].values[0]
    if label == '1FRdys':
        label = 1
    elif label == '2FRndys':
        label = 0
    elif label == '3Ctrl': #TODO: check if this is correct!
        label = 0
    labels.append([ID, label, label_risk, filenames])

In [29]:
labels[:10]

[['302', 0, 'notAtRisk_rest', ['302_17_mc_mmn25.cnt', '302_17_mc_mmn36.cnt']],
 ['170', 0, 'atRisk', ['170_17_mr_mmn25_wk.cnt', '170_17_mr_mmn_36_wk.cnt']],
 ['336', 0, 'notAtRisk_rest', ['336_17_mc_mmn36_wk.cnt']],
 ['027',
  0,
  'notAtRisk_rest',
  ['027_17_jc_mmn25.cnt',
   '027_17_jc_mmn25_2.cnt',
   '027_17_jc_mmn36.cnt',
   '027_17_jc_mmn47.cnt']],
 ['110', 1, 'notEnoughInfo', ['110_17_jr_mmn.cnt']],
 ['728', 'missing', 'atRisk', ['728-164-17m-jr-mmn.cnt']],
 ['637',
  0,
  'notAtRisk_highestScores',
  ['637-479-17m-mc-mmn36-2.cnt', '637-479-17m-mc-mmn36.cnt']],
 ['487',
  1,
  'notEnoughInfo',
  ['487_17_md_mmn_25_wk.cnt', '487_17_md_mmn_36_wk.cnt']],
 ['144',
  'missing',
  'notEnoughInfo',
  ['144_17_jr_mmn25_wk.cnt', '144_17_jr_mmn36.cnt']],
 ['010',
  '4',
  'notAtRisk_rest',
  ['010_17_jc_mmn36_wk_mmn25_wk_mmn47_wk_mmn58_wk.cnt']]]

In [66]:
len(labels), len(list(set(found_ids)))

(247, 247)

### Count number (and type) of labels found:

In [9]:
labels_known = 0
labels_unknown = 0
labels_type = []

for x in labels:
    if x[1] == 1: #'dyslexic'
        labels_known += 1
        labels_type.append(1)
    elif x[1] == 0: #'non-dyslexic'
        labels_known += 1
        labels_type.append(0)
    else: # missing or unclear
        labels_unknown += 1  
        labels_type.append('missing')
        
print("Data with proper labels:", labels_known, "||| Data without proper label:", labels_unknown)     

Data with proper labels: 209 ||| Data without proper label: 38


In [10]:
print("Data for 'dyslexic':", labels_type.count(1))
print("Data for 'non-dyslexic':", labels_type.count(0))

Data for 'dyslexic': 49
Data for 'non-dyslexic': 160


In [11]:
# Check types of risk group labels found
labels_risktype = [x[2] for x in labels]
list(set(labels_risktype))

['notAtRisk_highestScores',
 'notEnoughInfo',
 'notAtRisk_rest',
 'missing',
 'atRisk']

In [12]:
metadata['atRiskOrNotDDP'][:10]

0      unclear
1      missing
2    notAtRisk
3      missing
4    notAtRisk
5    notAtRisk
6    notAtRisk
7    notAtRisk
8    notAtRisk
9      unclear
Name: atRiskOrNotDDP, dtype: object

In [13]:
group_notrisk = np.array(1*((metadata['atRiskOrNotDDP'] == 'notAtRisk')
                   | (metadata['assignment4'].isin(['notAtRisk_rest', 'notAtRisk_highestScores']))))

group_risk = np.array(1*((metadata['atRiskOrNotDDP'] == 'atRisk')
                   | (metadata['assignment4'] == 'at risk')))

In [14]:
np.sum(group_risk) + np.sum(group_notrisk)

307

In [15]:
label_risk = group_notrisk + 2*group_risk 
label_risk[label_risk == 3] = 2
label_risk = label_risk -1

In [16]:
label_risk

array([-1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
        0,  0,  0,  0,  0,  0,  0,  0, -1,  0,  0,  0,  0, -1,  0,  0,  0,
        0,  0, -1, -1,  1, -1,  1,  0, -1,  0,  1,  1,  1, -1, -1,  0,  1,
        1,  0,  1,  1,  1,  1,  1,  1, -1,  1,  1,  1,  1, -1, -1,  1, -1,
        1,  1,  0,  1,  1,  1,  1,  1,  0, -1,  1,  1,  1,  1,  1, -1,  1,
        1,  1,  1,  1,  1,  1,  1,  1, -1,  1,  1,  1, -1,  1,  1,  1,  1,
        1,  1, -1,  1,  1, -1,  1,  1,  1,  1,  1,  1,  1,  0,  0,  0,  0,
        0, -1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
        0,  0,  0,  0,  0,  0,  0,  0,  0, -1,  0,  0,  0,  0,  0,  0, -1,
        0,  0,  0,  0,  0,  0, -1, -1,  1,  1, -1,  0, -1,  1,  1,  1,  1,
        1,  1,  0,  1,  1,  1,  1, -1, -1, -1,  1,  1,  1,  0,  1,  1,  1,
        1,  1,  1,  1, -1,  1,  1,  0, -1, -1,  1,  1,  1,  0,  1,  1,  1,
        1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1, -1,  1,  1,  1,  1,
        1,  1,  1,  1,  1

In [13]:
group_notdys = np.array(1*(metadata['groupDDP'].isin(['1FRdys', '3Ctrl'])))

group_dys = np.array(1*(metadata['groupDDP'] == '2FRndys'))

In [14]:
np.sum(group_notdys) + np.sum(group_dys)

252

In [15]:
label_dys = group_notdys + 2*group_dys 
label_dys[label_dys == 3] = 2
label_dys = label_dys -1

In [16]:
label_dys

array([-1, -1,  0, -1,  0,  0,  0,  0,  0, -1, -1, -1, -1,  0,  0,  0,  0,
        0,  0,  0, -1,  0,  0,  0,  0, -1,  0,  0,  0, -1, -1,  0,  0,  0,
        0,  0, -1, -1,  1, -1,  1, -1, -1, -1,  1,  1,  0, -1, -1, -1,  0,
        0, -1,  1,  1,  0,  0,  1,  1, -1,  0,  0,  1,  1, -1, -1,  1, -1,
        0,  1, -1,  1,  1,  0,  1,  0, -1, -1,  0,  1,  1,  0,  1, -1,  1,
        1,  1,  0,  1,  0,  1,  0,  1, -1,  1,  0,  1, -1,  0,  1,  0,  0,
        1,  0, -1,  0,  0, -1,  0,  1,  1,  1,  1,  1,  1,  0,  0, -1, -1,
        0, -1,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, -1,  0,  0,  0,  0,
        0,  0,  0, -1,  0,  0,  0, -1, -1, -1,  0,  0,  0,  0,  0,  0, -1,
       -1,  0,  0,  0,  0,  0, -1, -1,  1,  0, -1,  0, -1,  0,  0,  0,  0,
        0,  0, -1,  1,  1,  1,  1, -1, -1, -1,  1,  0,  1, -1,  1,  0,  0,
        1,  1,  0,  1, -1,  1,  0, -1, -1, -1,  1,  0,  1,  0,  1,  0,  1,
        1,  1,  1,  0,  0,  1,  0,  1,  1,  1,  0,  1, -1,  1,  1,  1,  0,
        1,  1,  0,  1,  1

## create Dataframe with labels to be used

In [24]:
labels_final = pd.DataFrame(data=metadata['id_child'].values, columns=['id_child'])
labels_final['label_dys'] = label_dys
labels_final['label_risk'] = label_risk
labels_final.head()

id_child  label_dys label_risk
0      001         -1     atRisk
1      002         -1     atRisk
2      003          0     atRisk
3      004         -1     atRisk
4      005          0     atRisk

In [25]:
print("Data for 'at risk':", labels_risktype.count('atRisk'))
print("Data for 'notAtRisk_rest':", labels_risktype.count('notAtRisk_rest'))
print("Data for 'notAtRisk_highestScores':", labels_risktype.count('notAtRisk_highestScores'))

Data for 'at risk': 91
Data for 'notAtRisk_rest': 56
Data for 'notAtRisk_highestScores': 49


In [26]:
labels_risktype = [x[2] for x in labels if x[1] in [1,0]]
print("Data for 'at risk':", labels_risktype.count('atRisk'))
print("Data for 'notAtRisk_rest':", labels_risktype.count('notAtRisk_rest'))
print("Data for 'notAtRisk_highestScores':", labels_risktype.count('notAtRisk_highestScores'))

Data for 'at risk': 79
Data for 'notAtRisk_rest': 43
Data for 'notAtRisk_highestScores': 45


In [27]:
metadata.loc[(metadata['groupDDP'].isin(['1FRdys', '2FRndys', '3Ctrl']) 
              & metadata['assignment4'].isin(['at risk', 'notAtRisk_rest', 'notAtRisk_highestScores', ]))]

id_child groupDDP atRiskOrNotDDP dyslexicAtMidGroup3DDP  \
2        003    3Ctrl      notAtRisk                      0   
4        005    3Ctrl      notAtRisk                      0   
5        006    3Ctrl      notAtRisk                      0   
8        009    3Ctrl      notAtRisk                      1   
15       016    3Ctrl      notAtRisk                      0   
16       017    3Ctrl      notAtRisk                      0   
17       018    3Ctrl      notAtRisk                      0   
19       021    3Ctrl      notAtRisk                      0   
21       023    3Ctrl      notAtRisk                      0   
22       024    3Ctrl      notAtRisk                      0   
23       025    3Ctrl      notAtRisk                      0   
24       027    3Ctrl      notAtRisk                      0   
26       029    3Ctrl      notAtRisk                      0   
31       034    3Ctrl      notAtRisk                      0   
32       035    3Ctrl      notAtRisk                      0   
35       039    3Ctrl      notAtRisk                      0   
50       114   1FRdys         atRisk                      1   
68       135   1FRdys         atRisk                      1   
72       139  2FRndys         atRisk                      0   
87       154   1FRdys         atRisk                      1   
89       156   1FRdys         atRisk                      1   
90       157  2FRndys         atRisk                      0   
96       164  2FRndys         atRisk                      0   
101      169   1FRdys         atRisk                      1   
109      177  2FRndys         atRisk                      1   
113      181  2FRndys         atRisk                      0   
115      301    3Ctrl      notAtRisk                      0   
116      302    3Ctrl      notAtRisk                      1   
119      305    3Ctrl      notAtRisk                      0   
121      307    3Ctrl      notAtRisk                      0   
..       ...      ...            ...                    ...   
235      608    3Ctrl      notAtRisk                      0   
236      609    3Ctrl      notAtRisk                      0   
237      610    3Ctrl      notAtRisk                      1   
238      611    3Ctrl      notAtRisk                      0   
240      613    3Ctrl      notAtRisk                      0   
241      614    3Ctrl      notAtRisk                      0   
242      615    3Ctrl      notAtRisk                      1   
243      616    3Ctrl      notAtRisk                      1   
244      617    3Ctrl      notAtRisk                      0   
245      618    3Ctrl      notAtRisk                      0   
246      619    3Ctrl      notAtRisk                      0   
247      620    3Ctrl      notAtRisk                      1   
249      622    3Ctrl      notAtRisk                      1   
251      625    3Ctrl      notAtRisk                      0   
252      626    3Ctrl      notAtRisk                      0   
253      627    3Ctrl      notAtRisk                      0   
254      628    3Ctrl      notAtRisk                      0   
255      629    3Ctrl      notAtRisk                      0   
258      632    3Ctrl      notAtRisk                      0   
259      633    3Ctrl      notAtRisk                      0   
263      637    3Ctrl      notAtRisk                      0   
265      639    3Ctrl      notAtRisk                      0   
266      640    3Ctrl      notAtRisk                      0   
267      641    3Ctrl      notAtRisk                      0   
268      642    3Ctrl      notAtRisk                      0   
271      646    3Ctrl      notAtRisk                      0   
272      701  2FRndys         atRisk                      0   
274      703   1FRdys         atRisk                      1   
309      738   1FRdys         atRisk                      1   
328      757  2FRndys         atRisk                      0   

                 assignment1              assignment2  \
2    notAtRisk_highestScores  notAtRisk_

## Custom cnt-file import function:

In [18]:
def read_cnt_file(file,
                  label_group,
                  event_idx = [3, 13, 66],
                  channel_set = "30",
                  tmin = -0.2,
                  tmax = 0.8,
                  lpass = 0.5, 
                  hpass = 40, 
                  threshold = 5, 
                  max_bad_fraction = 0.2):
    """ Function to read cnt file. Run bandpass filter. 
    Then detect and correct/remove bad channels and bad epochs.
    Store resulting epochs as arrays.
    
    Args:
    --------
    file: str
        Name of file to import.
    label_group: int
        Unique ID of specific group (must be >0).
    channel_set: str
        Select among pre-defined channel sets. Here: "30" or "62"
    """
    
    if channel_set == "30":
        channel_set = ['O2', 'O1', 'OZ', 'PZ', 'P4', 'CP4', 'P8', 'C4', 'TP8', 'T8', 'P7', 
                       'P3', 'CP3', 'CPZ', 'CZ', 'FC4', 'FT8', 'TP7', 'C3', 'FCZ', 'FZ', 
                       'F4', 'F8', 'T7', 'FT7', 'FC3', 'F3', 'FP2', 'F7', 'FP1']
    elif channel_set == "62":
        channel_set = ['O2', 'O1', 'OZ', 'PZ', 'P4', 'CP4', 'P8', 'C4', 'TP8', 'T8', 'P7', 
                       'P3', 'CP3', 'CPZ', 'CZ', 'FC4', 'FT8', 'TP7', 'C3', 'FCZ', 'FZ', 
                       'F4', 'F8', 'T7', 'FT7', 'FC3', 'F3', 'FP2', 'F7', 'FP1', 'AFZ', 'PO3', 
                       'P1', 'POZ', 'P2', 'PO4', 'CP2', 'P6', 'M1', 'CP6', 'C6', 'PO8', 'PO7', 
                       'P5', 'CP5', 'CP1', 'C1', 'C2', 'FC2', 'FC6', 'C5', 'FC1', 'F2', 'F6', 
                       'FC5', 'F1', 'AF4', 'AF8', 'F5', 'AF7', 'AF3', 'FPZ']
    else:
        print("Predefined channel set given by 'channel_set' not known...")
        
    
    # Initialize array
    signal_collection = np.zeros((0,len(channel_set),501))
    label_collection = np.zeros((0))
    
    # Import file 
    data_raw = mne.io.read_raw_cnt(file, montage=None, eog='auto', preload=True)
    
    # Band-pass filter (between 0.5 and 40 Hz. was 0.5 to 30Hz in Stober 2016)
    data_raw.filter(0.5, 40, fir_design='firwin')

    events = mne.find_events(data_raw, shortest_event=0, stim_channel='STI 014', verbose=False)
    
    # Set baseline:
    baseline = (None, 0)  # means from the first instant to t = 0

    # Select channels to exclude (if any)
    channels_exclude = [x for x in data_raw.ch_names if x not in channel_set]
    channels_exclude = [x for x in channels_exclude if x not in ['HEOG', 'VEOG', 'STI 014']]
    
    for event_id in event_idx:
    
        # Pick EEG channels 
        picks = mne.pick_types(data_raw.info, meg=False, eeg=True, stim=False, eog=False,
                           #exclude=data_exclude)#'bads'])
                               include=channel_set, exclude=channels_exclude)#'bads'])

        epochs = mne.Epochs(data_raw, events, event_id, tmin, tmax, proj=True, picks=picks,
                        baseline=baseline, preload=True, verbose=False)

        # Detect potential bad channels and epochs
        bad_channels, bad_epochs = helper_functions.select_bad_epochs(epochs, 
                                                                      event_id, 
                                                                      threshold = threshold, 
                                                                      max_bad_fraction = max_bad_fraction)

        # Interpolate bad channels
        if len(bad_channels) > 0: 
            # Mark bad channels:
            data_raw.info['bads'] = bad_channels
            # Pick EEG channels:
            picks = mne.pick_types(data_raw.info, meg=False, eeg=True, stim=False, eog=False,
                               #exclude=data_exclude)#'bads'])
                               include=channel_set, exclude=channels_exclude)#'bads'])
            epochs = mne.Epochs(data_raw, events, event_id, tmin, tmax, proj=True, picks=picks,
                            baseline=baseline, preload=True, verbose=False)
            # Interpolate bad channels using functionality of 'mne'
            epochs.interpolate_bads()

        # Get signals as array and add to total collection
        signals_cleaned = epochs[str(event_id)].drop(bad_epochs).get_data()
        signal_collection = np.concatenate((signal_collection, signals_cleaned), axis=0)
        #label_collection = np.concatenate((label_collection, event_id*label_group*np.ones((signals_cleaned.shape[0]))), axis=0)
        label_collection = [str(event_id) + label_group] * signals_cleaned.shape[0]

    return signal_collection, label_collection#.astype(int)

In [21]:
found_ids = [x[:3] for x in cnt_files]
idx = np.where(np.array(found_ids) == '036')[0]
[cnt_files[x] for x in idx]

['036_17_mc_mmn25_wk.cnt', '036_17_mc_mmn36.cnt']

In [22]:
path

NameError: name 'path' is not defined

In [73]:
# channel names for 30 EEG channel case: 
print(epochs.ch_names)

NameError: name 'epochs' is not defined

In [126]:
# channel names for 62 EEG channel case: 
print(epochs.ch_names)

['O2', 'O1', 'OZ', 'PZ', 'P4', 'CP4', 'P8', 'C4', 'TP8', 'T8', 'P7', 'P3', 'CP3', 'CPZ', 'CZ', 'FC4', 'FT8', 'TP7', 'C3', 'FCZ', 'FZ', 'F4', 'F8', 'T7', 'FT7', 'FC3', 'F3', 'FP2', 'F7', 'FP1', 'AFZ', 'PO3', 'P1', 'POZ', 'P2', 'PO4', 'CP2', 'P6', 'M1', 'CP6', 'C6', 'PO8', 'PO7', 'P5', 'CP5', 'CP1', 'C1', 'C2', 'FC2', 'FC6', 'C5', 'FC1', 'F2', 'F6', 'FC5', 'F1', 'AF4', 'AF8', 'F5', 'AF7', 'AF3', 'FPZ']


## Check how many EEG channels the cnt-files feature... 

In [30]:
format_collection = []
for i, filename in enumerate(cnt_files):
    # Import file 
    file = PATH_DATA + filename
    data_raw = mne.io.read_raw_cnt(file, montage=None, eog='auto', preload=True)
    format_collection.append((i, len(data_raw.ch_names)))
    print(i, len(data_raw.ch_names))

FileNotFoundError: [Errno 2] No such file or directory: 'C:\\OneDrive - Netherlands eScience Center\\Project_ePodium\\Data\\EEGdata_Karin_Jan2020\\001_17_jc_mmn36_slp_mmn25_slp_mmn47_slp_mmn58_slp.cnt'

In [111]:
a,b = zip(*format_collection)
len(np.where((np.array(b) == 65))[0]), len(np.where((np.array(b) == 33))[0]), len(a)

(128, 64, 192)

In [40]:
# number of cnt files which have both 65 channels (62 EEG) AND a label.
len(np.where((np.array(b) == 65) & (np.array(labels_type) == 1))[0])

NameError: name 'b' is not defined

So far we 'only' have about 60 cnt-files of which we have a label ("risk group" vs "no risc group").
And only 42 of them feature 62 EEG channels. I hence switched to 30 EEG channels and picked the ones that are present in all patient datasets.

# Workflow data processing
1. Load cnt files.
2. Select same number of channels (here: 30 same channels which exist for both 30 and 62 channel data)
3. Preprocess raw data (bandpass + detect outliers and 'bad' epochs).
4. Store epoch data and event type as array

## LABELS:
+ After Karin's search we have proper labels for much more files!  


In [28]:
PATH_CNTS = PATH_DATA + "17mnd mmn\\"

# Initialize array
signal_collection = np.zeros((0,30,501)) #62
label_collection = np.zeros((0))
metadata_collection = []

for i, filename in enumerate(cnt_files[:5]):
    
    # First check if we have proper label for that file
    # -----------------------------------------------------------
    
    ID = filename[:3]
    label = labels_final[labels_final['id_child'] == ID]['label_dys'].values[0]
    label_risk = labels_final[labels_final['id_child'] == ID]['label_risk'].values[0]
    #label = metadata[metadata['id_child'] == ID]['groupDDP'].values[0]
    #label_risk = metadata[metadata['id_child'] == ID]['assignment4'].values[0]
    
    if (label < 0) or (label_risk < 0):
        print("No proper label found for file: ", filename)
    else:
        #label_group = int(metadata[metadata["file"].str.match(filename[:-4])]['group'])
        label_group = 'dys' + str(label) + '_risk' + str(label_risk)
        
        print(40*"=")
        print("Importing file: ",filename)
        print("Data belongs into group: ", label_group)

        # Import data and events
        file = PATH_CNTS + filename

        signal_collect, label_collect = read_cnt_file(file, 
                                                      label_group,
                                                      event_idx = [3, 13, 66],
                                                      channel_set = "30",
                                                      tmin = -0.2,
                                                      tmax = 0.8,
                                                      lpass = 0.5, 
                                                      hpass = 40, 
                                                      threshold = 5, 
                                                      max_bad_fraction = 0.2)

        # Get signals as array and add to total collection
        signal_collection = np.concatenate((signal_collection, signal_collect), axis=0)
        label_collection = np.concatenate((label_collection, label_collect), axis=0)
        metadata_collection.append((i, filename, signal_collection.shape[0]))


No proper label found for file:  001_17_jc_mmn36_slp_mmn25_slp_mmn47_slp_mmn58_slp.cnt


TypeError: '<' not supported between instances of 'str' and 'int'

In [35]:
cnt_files[0]

'001_17_jc_mmn36_slp_mmn25_slp_mmn47_slp_mmn58_slp.cnt'

In [39]:
filename = cnt_files[0]
#data_raw = mne.io.read_raw_cnt(PATH_CNTS + filename, montage=None, eog='auto', preload=True)
data_raw = mne.io.read_raw_cnt(PATH_CNTS + filename, montage=None, eog='auto', preload=True)

Reading 0 ... 1500919  =      0.000 ...  3001.838 secs...


In [41]:
len(data_raw.ch_names)

64

In [42]:
data_raw.filter(0.5, 40, fir_design='firwin')

events = mne.find_events(data_raw, shortest_event=0, verbose=False)

Filtering raw data in 1 contiguous segment
Setting up band-pass filter from 0.5 - 40 Hz

FIR filter parameters
---------------------
Designing a one-pass, zero-phase, non-causal bandpass filter:
- Windowed time-domain design (firwin) method
- Hamming window with 0.0194 passband ripple and 53 dB stopband attenuation
- Lower passband edge: 0.50
- Lower transition bandwidth: 0.50 Hz (-6 dB cutoff frequency: 0.25 Hz)
- Upper passband edge: 40.00 Hz
- Upper transition bandwidth: 10.00 Hz (-6 dB cutoff frequency: 45.00 Hz)
- Filter length: 3301 samples (6.602 sec)



ValueError: No stim channels found, but the raw object has annotations. Consider using mne.events_from_annotations to convert these to events.

In [43]:
events = mne.events_from_annotations(data_raw)

Used Annotations descriptions: ['0', '12', '14', '15', '18', '2', '3', '4', '48', '5', '55', '66', '77', '80', '88']


In [53]:
signal_collection.shape, label_collection.shape

((39083, 30, 501), (39083,))

In [74]:
metadata_collection[:10]

[(1, '034_17_mc_mmn36_wk.cnt', 443),
 (2, '036_17_mc_mmn36_wk.cnt', 1384),
 (18, '175_17_jd_mmn_wk.cnt', 1858),
 (26, '305_17_jc_mmn36_wk.cnt', 2349),
 (27, '306_17_mc_mmn36_wk.cnt', 2813),
 (28, '307_17_jc_mmn36_wakker.cnt', 3447),
 (29, '308_17_jc_mmn36_wk.cnt', 3938),
 (30, '309_17_jc_mmn.cnt', 4425),
 (33, '314_17_mc_mmn36_wk.cnt', 4917),
 (38, '337_17_jc_mmn36_wk.cnt', 5389)]

We hence get a dataset of 39083 datapoints with known label.  
Each datapoint consits of a 1-second EEG signal of 30 channels with a 500Hz sampling rate. Thus arrays with a size of 30 x 501. 

## Labels 
Here we have 6 labels. 
1. Group 1, stimuli 3 --> "3"
2. Group 1, stimuli 13 --> "13"
3. Group 1, stimuli 66 --> "66"
4. Group 2, stimuli 3 --> "6"
5. Group 2, stimuli 13 --> "26"
6. Group 2, stimuli 66 --> "132"

In [41]:
label_collection[1500:2000].astype(int)

array([  6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,
         6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,
         6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,
         6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,
         6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,
         6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,
         6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,
         6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,
         6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,
         6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,
         6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,
         6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,
         6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,
         6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   6,   

# Save entire processed dataset:

In [75]:
filename = PATH_OUTPUT + "EEG_data_30channels_1s_corrected.npy"
np.save(filename, signal_collection)

filename = PATH_OUTPUT + "EEG_data_30channels_1s_corrected_labels.npy"
np.save(filename, label_collection)

import csv
filename = PATH_OUTPUT + "EEG_data_30channels_1s_corrected_metadata.csv"

with open(filename, 'w') as csvFile:
    writer = csv.writer(csvFile)
    writer.writerows(metadata_collection)
csvFile.close()